File: c++/src/services/agent/ChatGptMonitor.cpp
#include "ChatGptMonitor.hpp"

#include <QtConcurrent/QtConcurrentRun>

namespace Pet::Services {

ChatGptMonitor::ChatGptMonitor(const QString& configDir, QObject* parent)
    : QObject(parent)
    , m_configDir(configDir)
    , m_reader(std::make_unique<CodexStatusReader>())
    , m_timer(new QTimer(this)) {
    qRegisterMetaType<CodexSnapshot>("Pet::Services::CodexSnapshot");
    m_timer->setInterval(250);
    connect(m_timer, &QTimer::timeout, this, &ChatGptMonitor::poll);
}

ChatGptMonitor::~ChatGptMonitor() {
    disposeWorker();
}

void ChatGptMonitor::disposeWorker() {
    if (m_pool) {
        m_pool->clear(); // 丢弃尚未开始的任务
        m_pool.reset();  // 析构等待进行中的任务（有界的只读查询）
    }
    m_hasFuture = false;
    m_future = QFuture<CodexSnapshot>();
}

void ChatGptMonitor::start() {
    if (m_running) {
        return;
    }
    m_running = true;
    m_paused = false;
    m_generation++;
    m_pool = std::make_unique<QThreadPool>();
    m_pool->setMaxThreadCount(1); // 单个 worker：与 Python max_workers=1 等价
    m_timer->start(250);
    poll();
}

void ChatGptMonitor::stop() {
    m_running = false;
    m_paused = false;
    m_generation++;
    m_timer->stop();
    disposeWorker();
    m_snapshot = CodexSnapshot();
}

void ChatGptMonitor::pause() {
    if (m_running) {
        m_paused = true;
        m_generation++;
        m_timer->stop();
    }
}

void ChatGptMonitor::resume() {
    if (m_running && m_paused) {
        m_paused = false;
        m_generation++;
        m_timer->start(250);
    }
}

void ChatGptMonitor::poll() {
    if (!isRunning()) {
        return;
    }

    if (m_hasFuture) {
        if (!m_future.isFinished()) {
            return; // 上一轮仍在读，保持等待
        }
        CodexSnapshot snapshot;
        if (m_future.resultCount() > 0) {
            snapshot = m_future.result();
        } else {
            snapshot = CodexSnapshot{QStringLiteral("unavailable"), {}, {}, {}, {}, 0,
                                     QStringLiteral("监听暂不可用，稍后自动重试")};
        }
        m_hasFuture = false;
        m_future = QFuture<CodexSnapshot>();
        if (m_requestGeneration == m_generation) {
            publish(snapshot);
        }
    }

    if (m_pool) {
        m_requestGeneration = m_generation;
        CodexStatusReader* reader = m_reader.get();
        m_future = QtConcurrent::run(m_pool.get(), [reader]() {
            // CodexStatusReader 内部已把 sqlite/OS 异常折叠为「暂不可读」快照
            return reader->read();
        });
        m_hasFuture = true;
    }
}

void ChatGptMonitor::publish(const CodexSnapshot& snapshot) {
    const CodexSnapshot previous = m_snapshot;
    const bool changedTask =
        (previous.threadId != snapshot.threadId) || (previous.turnId != snapshot.turnId);

    m_snapshot = snapshot;

    const bool busy = snapshot.state == QLatin1String("thinking")
        || snapshot.state == QLatin1String("working")
        || snapshot.state == QLatin1String("attention");
    m_timer->setInterval(busy ? 250 : 1000);

    if (snapshot == previous) {
        return;
    }
    if (changedTask) {
        emit taskChanged();
    }
    emit snapshotChanged(snapshot);

    // 基线已完成任务与切换会话都不应庆祝旧历史：
    // 只有被跟踪回合上的真实状态迁移才算完成。
    if (!changedTask || busy || snapshot.state == QLatin1String("unavailable")) {
        emit stateChanged(m_agentKey, snapshot.state);
    }
    if (!snapshot.tool.isEmpty()
        && (snapshot.itemId != previous.itemId || snapshot.tool != previous.tool)) {
        emit activity(m_agentKey, snapshot.tool);
    }
}

} // namespace Pet::Services

File: python/pet/services/chatgpt_monitor.py
"""GUI-thread facade; local Codex reads run in one lazily created worker."""
from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor
from PySide6.QtCore import QObject, QTimer, Signal

from pet.infrastructure.codex_status import CodexSnapshot, CodexStatusReader


class ChatGPTMonitor(QObject):
    state_changed = Signal(str, str)
    activity = Signal(str, str)
    snapshot_changed = Signal(object)
    task_changed = Signal()

    def __init__(self, config_dir, parent=None, *, reader=None):
        super().__init__(parent)
        self.agent_key = 'chatgpt'
        self._reader = reader or CodexStatusReader()
        self._running = False
        self._paused = False
        self._executor = None
        self._future = None
        self._generation = 0
        self._request_generation = 0
        self.snapshot = CodexSnapshot()
        self._timer = QTimer(self)
        self._timer.setInterval(250)
        self._timer.timeout.connect(self._poll)
        # No callback from the worker touches this QObject after deletion.
        self._worker_holder = {}
        self.destroyed.connect(lambda *_, holder=self._worker_holder: _shutdown_worker(holder))

    def _dispose_worker(self, *_):
        if self._executor is not None:
            self._executor.shutdown(wait=False, cancel_futures=True)
            self._executor = None
            self._worker_holder.clear()

    def is_running(self):
        return self._running and not self._paused

    def start(self):
        if self._running:
            return
        self._running = True
        self._paused = False
        self._generation += 1
        self._executor = ThreadPoolExecutor(max_workers=1, thread_name_prefix='pet-codex-status')
        self._worker_holder['executor'] = self._executor
        self._timer.start(250)
        self._poll()

    def stop(self):
        self._running = False
        self._paused = False
        self._generation += 1
        self._timer.stop()
        self._dispose_worker()
        self._future = None
        self.snapshot = CodexSnapshot()

    def pause(self):
        if self._running:
            self._paused = True
            self._generation += 1
            self._timer.stop()

    def resume(self):
        if self._running and self._paused:
            self._paused = False
            self._generation += 1
            self._timer.start(250)

    def _poll(self):
        if not self.is_running():
            return
        if self._future is not None:
            if not self._future.done():
                return
            try:
                snapshot = self._future.result()
            except Exception:
                snapshot = CodexSnapshot(detail='监听暂不可用，稍后自动重试')
            self._future = None
            if self._request_generation == self._generation:
                self._publish(snapshot)
        if self._executor is not None:
            self._request_generation = self._generation
            self._future = self._executor.submit(self._reader.read)

    def _publish(self, snapshot):
        previous = self.snapshot
        changed_task = (previous.thread_id, previous.turn_id) != (snapshot.thread_id, snapshot.turn_id)
        self.snapshot = snapshot
        self._timer.setInterval(250 if snapshot.state in {'thinking','working','attention'} else 1000)
        if snapshot == previous:
            return
        if changed_task:
            self.task_changed.emit()
        self.snapshot_changed.emit(snapshot)
        # Baseline completed tasks and switching conversations must not celebrate
        # old history. Only a real transition on a tracked turn is a completion.
        if not changed_task or snapshot.state in {'thinking','working','attention','unavailable'}:
            self.state_changed.emit('chatgpt', snapshot.state)
        if snapshot.tool and (snapshot.item_id, snapshot.tool) != (previous.item_id, previous.tool):
            self.activity.emit('chatgpt', snapshot.tool)


def _shutdown_worker(holder):
    executor = holder.pop('executor', None)
    if executor is not None:
        executor.shutdown(wait=False, cancel_futures=True)
